### ISSUE TITLE
vkDeviceWaitIdle: returned VK_ERROR_DEVICE_LOST after loop 666

### ISSUE THREAD

**Issue by davidhubbard** [davidhubbard/v0lum3#1]
X server freezes up and application hangs. This might be a resource leak, since it is affected by how many times the main loop runs.

See the code in main/main.cpp:

```
    fprintf(stderr, "vkQueuePresentKHR done, end of loop %d\n", count);
    //
    // Probability of vkDeviceWaitIdle() failing with error:
    // "W ParameterValidation: code9: vkDeviceWaitIdle: returned VK_ERROR_DEVICE_LOST, indicating
    // that the logical device has been lost"
    // "vkDeviceWaitIdle returned -4"
    // 0 <= count <= 664: 0%
    // 0 <= count <= 665: 5%
    // 0 <= count <= 666: 100%
    //
    count++;
    if (count > 666) break;
```

The relevant output is:

```
vkQueuePresentKHR done, end of loop 665
vkAcquireNextImageKHR
vkQueueSubmit
D MEM: code0: Details of Memory Object list (of size 0 elements)
D MEM: code0: =============================
D MEM: code0: Details of CB list (of size 3 elements)
D MEM: code0: ==================
D MEM: code0:     CB Info (0x0x2581dd0) has CB 0x0x2588720
D MEM: code0:     CB Info (0x0x257ecf0) has CB 0x0x2587900
D MEM: code0:     CB Info (0x0x2592be0) has CB 0x0x257f5d0
vkQueuePresentKHR
vkQueuePresentKHR done, end of loop 666
vkDeviceWaitIdle
```

At this point, the X server hangs (the mouse cursor stops responding), the app hangs, and after ~2 seconds the app resumes with:

```
W ParameterValidation: code9: vkDeviceWaitIdle: returned VK_ERROR_DEVICE_LOST, indicating that the logical device has been lost
vkDeviceWaitIdle returned -4
```

nvidia-smi output:

```
$ nvidia-smi
Wed Aug  3 15:50:44 2016       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.35                 Driver Version: 367.35                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro K600         Off  | 0000:04:00.0      On |                  N/A |
| 25%   51C    P8    N/A /  N/A |    183MiB /   979MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:05:00.0     Off |                  N/A |
| 22%   36C    P8    15W / 250W |      1MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      3817    G   /usr/bin/X                                      96MiB |
|    0      5543    G   ...iveTaskBlocking/Enabled/StrictSecureCooki    86MiB |
+-----------------------------------------------------------------------------+
```

**Comment by davidhubbard** [comment]
Commit e26515b contains the test case. To build it:
1. Clone this repository
2. Run `build.sh` - it will ask you to `export PKG_CONFIG_PATH=$PWD/vendor/lib/pkgconfig` and then re-run `build.sh`
3. Run `make`
4. Run `main/v`

**Comment by davidhubbard** [comment]
I have posted on [forums.khronos.org](https://forums.khronos.org/showthread.php/13144-vulkan-tutorial-com-type-app-which-hangs-on-linux-nVidia-K600) for assistance.

**Comment by krOoze** [comment]
Your subpass dependencies seem wrong to me. There is missing dependency from src=0 to dst=EXTERNAL.

**Comment by davidhubbard** [comment]
Good idea! If this turns out to be the case, then vulkan-tutorial.com needs to be updated. I am using a refactored version of this code:

https://vulkan-tutorial.com/Drawing_a_triangle/Swap_chain_recreation

**Comment by davidhubbard** [comment]
I hit the "Comment" button too soon. This is the source code: https://vulkan-tutorial.com/code/swap_chain_recreation.cpp

If you search for VkSubpassDependency you will notice that only the dependency from src=EXTERNAL to dst=0 is added. Sascha Willems' triangle demo does not add any subpass dependencies (in vulkanexamplebase.cpp, renderPassInfo.dependencyCount = 0;)

**Comment by davidhubbard** [comment]
I will be on business-related travel until Aug 15 and will not respond to messages.

**Comment by krOoze** [comment]
Sascha Willems uses Pipeline Barriers instead for that purpose...

**Comment by krOoze** [comment]
Also your `srcStage` of the first dependency is `STAGE_BOTTOM`. But I think it should be `STAGE_COLOR_ATTACHMENT_OUTPUT` (i.e. matching the waitStage of your `SubmitInfo`).

**Comment by davidhubbard** [comment]
The issue was resolved by upgrading my nvidia driver from 367.35 to 370.23.

If anyone is still interested in investigating, commit 4c0ecee was the test case (be sure to use linux nvidia driver 367.35 as well).
